### PR TITLE
[cli] add dom components support to export command

### PR DESCRIPTION
**Pull request by Kudo**
# Why

added eas updates support to dom components
close ENG-12926

# How

- refactor common code for `export` and `export:embed`  to `exportDomComponents.ts`
- since updates serve assets in flattened MD5 based names, we cannot export web files like `export:embed`. this pr tries to post-transform bundles to have flattened MD5 based names.
- ❌ since updates serves in flattened files, we cannot support [public assets](https://docs.expo.dev/guides/dom-components/#public-assets) for updates. will have a pr to document that later.

# Test Plan

- add an unit test for `updateDomComponentAssetsForMD5Naming`
- router-e2e dom components debug build
- router-e2e dom components release build
- test eas updates in a default template project

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by Kudo** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/expo/expo/32504?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#32504** <a href="https://app.graphite.dev/github/pr/expo/expo/32504?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#32502** <a href="https://app.graphite.dev/github/pr/expo/expo/32502?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Kudo and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by arasrezaei** [comment]
this is huge!

File: packages/expo/src/dom/base.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    const localAssets = ExpoUpdates?.localAssets ?? {};
+    const anyLocalAsset = Object.values(localAssets)[0];
+    if (anyLocalAsset) {
+      // Try to get the `.expo-internal` directory from the first local asset
+      cachedBaseUrl = anyLocalAsset.slice(0, anyLocalAsset.lastIndexOf('/'));
+      return cachedBaseUrl;
+    }
+  }
+
+  if (process.env.EXPO_OS === 'web') {
```
Comment (EvanBacon) [inline packages/expo/src/dom/base.ts:29]: Web case shouldn't matter right? All content is just passed through.

File: packages/expo/src/dom/base.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    return cachedBaseUrl;
+  }
+
+  // Serving from updates
+  const ExpoUpdates = globalThis.expo?.modules?.['ExpoUpdates'] as
+    | import('expo-updates').ExpoUpdatesModule
+    | undefined;
+  const updatesIsInstalledAndEnabled = ExpoUpdates?.isEnabled ?? false;
+  const updatesIsEmbeddedLaunch = ExpoUpdates?.isEmbeddedLaunch ?? false;
+  const shouldServeDomFromUpdates = updatesIsInstalledAndEnabled && !updatesIsEmbeddedLaunch;
```
Comment (EvanBacon) [inline packages/expo/src/dom/base.ts:17]: Refactor into different function

File: packages/expo/src/dom/base.ts (hunk 1 above)
Comment (Kudo) [reply packages/expo/src/dom/base.ts:13]: yes, on web, the call path should not go to `webview-wrapper.tsx -> base.ts`. the web code here currently is only used for unit test

File: packages/expo/src/dom/base.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    } else {
+      cachedBaseUrl = process.env.EXPO_BASE_URL ?? '';
+    }
+    return cachedBaseUrl;
+  }
+
+  // Serving from local dev server
+  const getDevServer = require('react-native/Libraries/Core/Devtools/getDevServer');
+  const devServer = getDevServer();
+  cachedBaseUrl = new URL('/_expo/@dom', devServer.url).toString();
```
[the remark this answers is not in the document]
Comment (Kudo) [reply packages/expo/src/dom/base.ts:33]: it's baseurl, the final url will later append `file.tsx` in https://github.com/expo/expo/blob/37e96aa9a301e6fa501b24892dd4021adc09fd8a/packages/expo/src/dom/webview-wrapper.tsx#L54

**Comment by Kudo** [comment]
@EvanBacon  sorry that i did not merge #32502 through graphite website, so it did not rebase the pr automatically and accidentally having changes from #32502. i've rebased again so that the pr should be more cli specific.

**Review by EvanBacon — APPROVED** [review]
hard to fully follow this but overall seems good

**Pull request merged**